warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,533 @@
+    }
+
+    if (!webAccessController.checkURL(url)) {
+      throw new WarpScriptException(getName() + " invalid host or scheme in URL.");
+    }
+
+    //
+    // Check that we do not exceed the limits
+    //
+
+    // Get the current counters in the stack and initialize them if not present.
+    AtomicLong urlCount;
+    AtomicLong downloadSize;
+
+    Object ufCount = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_REQUESTS);
+    Object ufSize = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE);
+
+    if (null == ufCount || null == ufSize) {
+      urlCount = new AtomicLong();
+      downloadSize = new AtomicLong();
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_REQUESTS, urlCount);
+      stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE, downloadSize);
+    } else {
+      urlCount = (AtomicLong) ufCount;
+      downloadSize = (AtomicLong) ufSize;
+    }
+
+    if (urlCount.addAndGet(1) > maxrequests) {
+      throw new WarpScriptException(getName() + " is limited to " + maxrequests + " calls per script execution.");
+    }
+
+    Map<String, Object> res = new LinkedHashMap<>();
+    HttpURLConnection conn = null;
+
+    try {
+      conn = (HttpURLConnection) url.openConnection();
+
+      //
+      // Set headers
+      //
+
+      if (null == headersMacro) {
+
+        Object username = params.get(USERNAME);
+        Object password = params.get(PASSWORD);
+
+        if (null != username && null != password) {
+
+          //
+          // Compute basic auth
+          //
+
+          if (!(username instanceof String)) {
+            throw new WarpScriptException(getName() + " expects a STRING username when doing basic authentication.");
+          }
+
+          if (!(password instanceof String)) {
+            throw new WarpScriptException(getName() + " expects a STRING password when doing basic authentication.");
+          }
+
+          String userInfo = ((String) username) + ":" + ((String) password);
+          String basicAuth = "Basic " + Base64.encodeBase64String(userInfo.getBytes(StandardCharsets.UTF_8));
+          conn.setRequestProperty("Authorization", basicAuth);
+        }
+
+      } else {
+
+        stack.push(params);
+        stack.exec(headersMacro);
+        o = stack.pop();
+
+        if (!(o instanceof Map)) {
+          throw new WarpScriptException(getName() + " expects the " + HEADERS_MACRO + " to push a headers map onto the stack.");
+        }
+
+        headers = (Map) o;
+      }
+
+      for (Map.Entry<Object, Object> prop: headers.entrySet()) {
+        conn.setRequestProperty(String.valueOf(prop.getKey()), String.valueOf(prop.getValue()));
+      }
+
+      conn.setDoInput(true);
+      conn.setRequestMethod(method.toUpperCase());
+
+      //
+      // Set body
+      //
+
+      if ("GET" != method && "DELETE" != method && "TRACE" != method && "OPTIONS" != method && "HEAD" != method) {
```

Comment:
In fact, only TRACE cannot have a body, see the source code of [HttpURLConnection](https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L1319). So either check for TRACE (using `equals`, not `==`) or handle the IOException it may throw.

The method is changed from GET to POST if there is a body by the getOutputStream. My opinion is to let that happen. This is coherent with defaulting to "POST" is no method is defined and there is a body.

Reply:
I think it can be confusing if a method is explicit in the arguments and it gets changed to POST.

Reply:
You're right. It that case, just check for `TRACE` and `GET`.